hedera-node/docs/design/services/smart-contract-service/contract-accounts-nonces.md

```diff
@@ -34,43 +25,64 @@ The following is a table with general use cases and behavior for Ethereum and He
-- This would be done via setting it inside it's `ExpirableTxnRecord` in the corresponding `EvmFnResult` field
-    - `contractCreateResult` if the top level transaction was `ContractCreate`
-    - `contractCallResult` if the top level transaction was `ContractCall`
-
-### Migration Of Nonces For Existing Contracts
-
-- We need perform a migration for all existing contract nonces from Consensus Node state to Mirror Node e.g. similar to traceability migration performed in the past
-- For all contracts in state we should produce synthetic transaction records that would contain the contract id and it's corresponding nonce value
-  - We will create a synthetic `ContractCall` transaction body with:
-    - `contractID` field set to the contract which nonce we want to externalize
-    - `gas` field set to 0
-    - `amount` field set to 0
-    - `functionParameters` field set to empty byte array
-  - For each synthetic `ContractCall` transaction we will create a corresponding record with the `contractCallResult` field set to `ContractFunctionResult` value containing the following:
-    - `contractID` field set to the contract which nonce we want to externalize
-    - `contractCallResult` field set to empty byte array
-    - `errorMessage` field set to empty string
-    - `bloom` field set to empty byte array
-    - `gasUsed` field set to 0
-    - `contract_nonces` field set to `ContractNonceInfo` value with `contract_id` field set to the contract which nonce we want to externalize and `nonce` field set to the contract nonce value
+- Contract account nonces are updated in Consensus Node state using in the following places:
+  - [HederaWorldState](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/HederaWorldState.java)
+  - [ContractCreateTransitionLogic](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/contract/ContractCreateTransitionLogic.java#L209)
+  - [ContractCallTransitionLogic](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/contract/ContractCallTransitionLogic.java#L148)
+  - [TransactionProcessingResult](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/execution/TransactionProcessingResult.java)
+  - [EvmFnResult](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/submerkle/EvmFnResult.java)
+
+
+- We keep a `ContractId -> nonce` tree map inside `HederaWorldState`, it is updated on each call of `commit()` (using newly added method `trackContractNonces()`).
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**suggestion:** ‏Let's add the above links into the proper places in this description.